### DISCUSSION TITLE
DataGridColumn with null field exception

### DISCUSSION THREAD

**Discussion by sturlath** [Megabit/Blazorise#2842] [General]
Hi

I have a multitenant system that reviles certain information if you are a tenant or not. 

So in one case I want to display information in  my host count that will have null values in them.

When its null, I will display "n/a" or something.

But I´m having problems with this here where `EventPriceInfo` on `EventWithNavigationPropertiesDto` in `UpcomingEventList` is null **and therefore** `Field="EventPriceInfo.LivePrice"` throws

``` xml
  <DataGrid TItem="EventWithNavigationPropertiesDto"
        Data="UpcomingEventList"
        TotalItems="TotalCount"
        ShowPager="true"
        Responsive="true"
        PageSize="PageSize">
      <DataGridColumns>
         <DataGridColumn TItem="EventWithNavigationPropertiesDto"
                    Field="EventPriceInfo.LivePrice"
                    Caption="Price">
        </DataGridColumn>
      </DataGridColumns>
  </DataGrid>
```  
with this code-behind
```csharp
private IReadOnlyList<EventWithNavigationPropertiesDto> UpcomingEventList { get; set; } 
```

```csharp
    public class EventWithNavigationPropertiesDto
    {
        public EventPriceInfoDto EventPriceInfo { get; set; }
    }

    public class EventPriceInfoDto : FullAuditedEntityDto<Guid>
    {
        public int LivePrice { get; set; }
    }
```

The error (btw was hard to find out what was causing this in the first place. Is it possible to add more data?)

> Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
       Unhandled exception rendering component: Object reference not set to an instance of an object.
 System.NullReferenceException: Object reference not set to an instance of an object.
    at Blazorise.DataGrid.DataGridColumn`1[[EventWithNavigationPropertiesDto, Application.Contracts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].GetValue(StreamEventWithNavigationPropertiesDto item)
    at Blazorise.DataGrid.DataGridColumn`1[[EventWithNavigationPropertiesDto, Application.Contracts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].FormatDisplayValue(EventWithNavigationPropertiesDto item)
    at Blazorise.DataGrid._DataGridRow`1.<>c__DisplayClass0_0[[EventWithNavigationPropertiesDto, .Application.Contracts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].<BuildRenderTree>b__2(RenderTreeBuilder __builder3)
    at Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder.AddContent(Int32 sequence, RenderFragment fragment)
    at Blazorise.TableRowCell.BuildRenderTree(RenderTreeBuilder __builder)
    at Microsoft.AspNetCore.Components.ComponentBase.<.ctor>b__6_0(RenderTreeBuilder builder)
    at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment)
    at Microsoft.AspNetCore.Components.RenderTree.Renderer.RenderInExistingBatch(RenderQueueEntry renderQueueEntry)
    at Microsoft.AspNetCore.Components.RenderTree.Renderer.ProcessRenderQueue()

**Comment by stsrki** [comment]
Is this a typo? The field name is `LiveStreamPrice`, but you defined it as `LivePrice`(EventPriceInfo.LivePrice).

**Reply by sturlath** [reply]
No, sorry that was just a typo. I was just simplifying the code (there is lots of other columns in there). 
If I remove this column code there is no error. And if its not null and there is value in LivePrice it just works fine.

So **still** an issue :-)

**Comment by stsrki** [comment]
@David-Moreira Can you look into this, please?

**Reply by David-Moreira** [reply]
Hello,

code to reproduce:
```
<DataGrid TItem="EventWithNavigationPropertiesDto"
          Data="UpcomingEventList"
          TotalItems="5"
          ShowPager="true"
          Responsive="true"
          PageSize="10">
    <DataGridColumns>
        <DataGridColumn TItem="EventWithNavigationPropertiesDto"
                        Field="EventPriceInfo.LivePrice"
                        Caption="Price">
        </DataGridColumn>
    </DataGridColumns>
</DataGrid>

@code
{
    private IReadOnlyList<EventWithNavigationPropertiesDto> UpcomingEventList { get; set; } = new List<EventWithNavigationPropertiesDto>() { new() { } };


    public class EventWithNavigationPropertiesDto
    {
        public EventPriceInfoDto EventPriceInfo { get; set; }
    }

    public class EventPriceInfoDto
    {
        public int LivePrice { get; set; }
    }
}
```

Relevant documentation : 
https://blazorise.com/docs/extensions/datagrid/#nested-fields

> Just keep in mind that when **editing nested fields they must be initialized first** or otherwise they will raise an exception.

So on this case, we are not editting but we're just trying to display the field. and it still raises an exception. Thing is, Blazorise internally has to run a getter to do some setup, mainly in :
![Get](https://user-images.githubusercontent.com/22283161/131586178-60971f8b-69b4-4958-8495-12f8ad5ae868.png)
And 
![Format](https://user-images.githubusercontent.com/22283161/131586256-2f17ad7d-418d-4b2b-ba49-1dd4865d91ae.png)

@sturlath So for now there's no other solution then to just actually initialize the property.

@stsrki with the addition of the PR #2829 which will help with the editing of nested fields, we could potentially also take a look at allowing this? Attempt for 0.9.5?
Still has to be tought out... do we allow a full nested tree of null properties? What do we do, when we find that the parent of a property is null? We display the `default` of that property type? In the case of the OP's posted issue, it would be a 0.
So what do you think?

**Reply by stsrki** [reply]
I believe for field getter we should allow for nullable fields. I remember I have created a `safe` field accessor to allow for this.

![image](https://user-images.githubusercontent.com/900302/131635958-4214f150-aec7-49ac-9aa5-0a030a036e91.png)

If it doesn't work that means we accidentally break it at some point.

**Reply by David-Moreira** [reply]
Strange... Gotta take a look at the history... Are you sure that was ever working? haha maybe the intention was there but it never did because from looking at the code I think you'd need to compile quite a different expression and if I remember well... This method body mostly looked like this, and has had that IsNullable check, since I've started helping with Blazorise which should be something different then checking for null...
Eithereway I'll take a look later.

**Reply by stsrki** [reply]
I might be wrong. Can't remember everything 😅

**Reply by David-Moreira** [reply]
There were some iterations but since the first commit it's like this:
![image](https://user-images.githubusercontent.com/22283161/131749331-6b9d1b3c-f18c-4b68-81da-dd98d8f9bc7f.png)
So basically if the property we are accessing is `Nullable` value type, it handles it appropriately. 

Here, the use case @sturlath exposed, might seem easy at first glance, but it's a bit more complex. 

The null check isn't at the end of the traversal, as it is for a `Nullable<T> where T : Struct`, but it is as we are navigating the property/field "tree" constantly checking for a null reference type. And actually having to actually return a valid value for the property we are accessing at the end of the traversal, which will most likelly be a ValueType and need to be set to it's default.

Basically having to build the expression correspondent to this type of null coalescing:
`Parent1?.Parent2?.Parent3?.PropertyIWantToAccess`

Since that wasn't initialy supported I would attempt this for 0.9.5 so we don't mess with the "expression compiler" in 0.9.4.

**Reply by stsrki** [reply]
I agree. We need to do this for 0.9.5.

**Reply by sturlath** [reply]
@stsrki just to follow up on this. Will this be fixed for 0.9.5?

**Reply by stsrki** [reply]
Yes, I believe @David-Moreira already fixed it.

**Reply by sturlath** [reply]
Found it https://github.com/Megabit/Blazorise/issues/2855
